aspnet/fundamentals/configuration.rst

```diff
@@ -2,169 +2,252 @@
+  :language: json
+
+With the following ``Controller``:
+
+.. literalinclude:: configuration/sample/src/UsingOptions/Controllers/HomeController2.cs
+  :language: none
+  :start-after: snippet1
+  :end-before: #endregion
   :dedent: 4
-  :emphasize-lines: 3,5,8
 
-.. tip:: Learn more about :doc:`dependency-injection`.
+``subOption1 = subvalue1_from_json, subOption2 = 200`` is returned.
 
-To setup the :dn:iface:`~Microsoft.Extensions.Options.IOptions\<TOptions>` service you call the :dn:method:`~Microsoft.Extensions.DependencyInjection.OptionsServiceCollectionExtensions.AddOptions` extension method during startup in your ``ConfigureServices`` method:
+.. You can create a custom :dn:iface:`~Microsoft.Extensions.Options.IConfigureOptions\<TOptions>` service  (for example, to read settings from a database).  Use ``AddSingleton<IConfigureOptions<TOptions>>`` extension method to register a custom service.
 
-.. literalinclude:: configuration/sample/src/UsingOptions/Startup.cs
+.. _in-memory-provider:
+
+In-memory provider and binding to a POCO class
+------------------------------------------------
+
+The following sample shows how to use the in-memory provider and bind to a class:
+
+.. literalinclude:: configuration/sample/src/InMemory/Program.cs
+  :language: none
+
+.. literalinclude:: configuration/sample/src/InMemory/MyWindow.cs
+  :language: none
+
+Note the ConfigurationBinder’s ``GetValue<T>`` extension method allows you to specify a default value (80 in the sample)::
+
+   var left = Configuration.GetValue<int>("AppConfiguration:MainWindow:Left", 80);
+
+Configuration values are not limited to scalars. You can retrieve POCO objects or even entire object graphs. The following sample shows how to bind to the ``MyWindow`` class and use the options pattern with a ASP.NET Core MVC app:
+
+.. literalinclude:: configuration/sample/src/WebConfigBind/MyWindow.cs
   :language: c#
-  :lines: 26-30
-  :emphasize-lines: 4
+
+.. literalinclude:: configuration/sample/src/WebConfigBind/appsettings.json
+  :language: json
+
+Bind the custom class in ``ConfigureServices`` in the ``Startup`` class:
+
+.. literalinclude:: configuration/sample/src/WebConfigBind/Startup.cs
+  :language: none
   :dedent: 8
+  :start-after: snippet1
+  :end-before: #endregion
 
-.. _options-example:
+Display the settings from the ``HomeController``:
 
-The ``Index`` view displays the configured options:
+.. literalinclude:: configuration/sample/src/WebConfigBind/Controllers/HomeController.cs
+  :language: none
 
-.. image:: configuration/_static/index-view.png
+.. _custom-config-providers:
 
-You configure options using the :dn:method:`~Microsoft.Extensions.DependencyInjection.OptionsServiceCollectionExtensions.Configure\<TOptions>` extension method. You can configure options using a delegate or by binding your options to configuration:
+Note that Dependency Injection (DI) is not setup until after ``ConfigureServices`` is invoked and the configuration system is not DI aware.
```

Comment:
moved to bottom.  This is a copy/paste from Dan.Roth, folks that need to know will figure it out.